src/connectors/impls/gcl/writer/sink.rs

```diff
@@ -12,25 +12,222 @@
 use crate::connectors::prelude::*;
 use crate::connectors::utils::pb;
+use async_std::channel::bounded;
+use async_std::channel::Sender;
 use async_std::prelude::FutureExt;
+use async_std::task::JoinHandle;
 use googapis::google::logging::v2::log_entry::Payload;
 use googapis::google::logging::v2::logging_service_v2_client::LoggingServiceV2Client;
 use googapis::google::logging::v2::{LogEntry, WriteLogEntriesRequest};
 use gouth::Token;
 use prost_types::Timestamp;
-use std::time::Duration;
 use tonic::codegen::InterceptedService;
 use tonic::transport::{Certificate, Channel, ClientTlsConfig};
 use tonic::{Code, Status};
+use tremor_common::time::nanotime;
+
+#[allow(clippy::large_enum_variant)]
+pub(crate) enum GclEvent {
+    #[allow(dead_code)] // NOTE This is used by the tests and not intended for use in production
+    Shutdown,
+    LogEvent(Event, Vec<LogEntry>),
+}
+
+impl GclEvent {
+    async fn log_event(config: &Config, event: Event) -> Result<GclEvent> {
+        let timestamp = event.ingest_ns;
+        let mut entries = Vec::with_capacity(event.len());
+        for (data, meta) in event.value_meta_iter() {
+            let meta = meta.get("gcl_writer").or(None);
+            #[allow(clippy::cast_precision_loss)]
+            #[allow(clippy::cast_possible_wrap)]
+            let mut timestamp = Timestamp {
+                seconds: (timestamp / 1_000_000_000) as i64,
+                nanos: (timestamp % 1_000_000_000) as i32,
+            };
+            timestamp.normalize();
+            entries.push(value_to_log_entry(timestamp, config, data, meta)?);
+        }
+
+        Ok(GclEvent::LogEvent(event, entries))
+    }
+}
+
+pub(crate) struct GclClient {
+    #[allow(unused)] // NOTE FIXME consider kill switch and quiescence
+    join_handle: JoinHandle<Result<()>>,
+    #[allow(unused)] // NOTE Preserved so that auth token can be refreshed and renewed
+    token: Token,
+    sender: Sender<GclEvent>,
+}
+
+pub(crate) struct GclClients {
+    client_senders: Vec<GclClient>,
+    idx: usize,
+}
+
+impl GclClients {
+    // FIXME This will need to be refactored for fake/mock unit testing
+    async fn make_client(
+        config: &Config,
+        token: &Token,
+    ) -> std::result::Result<
+        LoggingServiceV2Client<InterceptedService<Channel, AuthInterceptor>>,
+        Error,
+    > {
+        let tls_config = ClientTlsConfig::new()
+            .ca_certificate(Certificate::from_pem(googapis::CERTIFICATES))
+            .domain_name("logging.googleapis.com");
+
+        let channel = Channel::from_static("https://logging.googleapis.com")
+            .connect_timeout(Duration::from_nanos(config.connect_timeout))
+            .tls_config(tls_config)?
+            .keep_alive_while_idle(true)
+            .tcp_keepalive(Some(Duration::from_secs(10)))
+            .http2_keep_alive_interval(Duration::from_secs(10))
+            .connect()
+            .await?;
+
+        let token_header_value = token.header_value();
```

Comment:
this function needs to be called in the interceptor function, as this is the place where the token is renewed.